tachyon/math/polynomials/multivariate/multilinear_dense_evaluations.h

```diff
@@ -0,0 +1,131 @@
+// Copyright 2022 arkworks contributors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.arkworks and the LICENCE-APACHE.arkworks
+// file.
+
+#ifndef TACHYON_MATH_POLYNOMIALS_MULTIVARIATE_MULTILINEAR_DENSE_EVALUATIONS_H_
+#define TACHYON_MATH_POLYNOMIALS_MULTIVARIATE_MULTILINEAR_DENSE_EVALUATIONS_H_
+
+#include <stddef.h>
+
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "tachyon/base/bits.h"
+#include "tachyon/base/containers/container_util.h"
+#include "tachyon/base/logging.h"
+#include "tachyon/base/strings/string_util.h"
+#include "tachyon/math/polynomials/multivariate/support_poly_operators.h"
+
+namespace tachyon::math {
+
+template <typename F, size_t MaxDegree>
+class MultilinearDenseEvaluations {
+ public:
+  constexpr static const size_t kMaxDegree = MaxDegree;
+
+  using Field = F;
+
+  constexpr MultilinearDenseEvaluations()
+      : MultilinearDenseEvaluations({F::Zero()}) {}
+  constexpr explicit MultilinearDenseEvaluations(
+      const std::vector<F>& evaluations)
+      : evaluations_(evaluations) {
+    CHECK_LE(Degree(), MaxDegree);
+  }
+  constexpr explicit MultilinearDenseEvaluations(std::vector<F>&& evaluations)
+      : evaluations_(std::move(evaluations)) {
+    CHECK_LE(Degree(), MaxDegree);
+  }
+
+  constexpr static MultilinearDenseEvaluations Zero(size_t degree) {
+    return MultilinearDenseEvaluations(
+        base::CreateVector(size_t{1} << degree, F::Zero()));
+  }
+
+  constexpr static MultilinearDenseEvaluations One(size_t degree) {
+    return MultilinearDenseEvaluations(
+        base::CreateVector(size_t{1} << degree, F::One()));
+  }
+
+  constexpr static MultilinearDenseEvaluations Random(size_t degree) {
+    return MultilinearDenseEvaluations(
+        base::CreateVector(size_t{1} << degree, []() { return F::Random(); }));
+  }
+
+  constexpr bool operator==(const MultilinearDenseEvaluations& other) const {
+    return evaluations_ == other.evaluations_;
+  }
+
+  constexpr bool operator!=(const MultilinearDenseEvaluations& other) const {
+    return !operator==(other);
+  }
+
+  constexpr F* Get(size_t i) {
+    return const_cast<F*>(std::as_const(*this).Get(i));
+  }
+
+  constexpr const F* Get(size_t i) const {
+    if (i < evaluations_.size()) {
+      return &evaluations_[i];
+    }
+    return nullptr;
+  }
+
+  constexpr bool IsZero() const {
+    return std::all_of(evaluations_.begin(), evaluations_.end(),
+                       [](const F& value) { return value.IsZero(); });
+  }
+
+  constexpr bool IsOne() const {
+    return std::all_of(evaluations_.begin(), evaluations_.end(),
+                       [](const F& value) { return value.IsOne(); });
+  }
+
+  constexpr size_t Degree() const {
+    return base::bits::SafeLog2Ceiling(evaluations_.size());
+  }
+
```

Comment:
```suggestion
  // The two-variate multilinear polynomial x₀ + 3 * x₀ * x₁ + 2 evaluates to
  // [2, 3, 2, 6] in the two-dimensional hypercube with points [00, 10, 01, 11].
  // which means let poly = [2, 3, 2, 6], partial_point = [1, 3].
  // then poly.Evaluate(partial_point) = 1 + 3 * 1 * 3 + 2 = 5 in GF7.
  // This can be confirmed in the unittest.
  // If the dimension is more then 3, It's convenient to flip the order of bit
  // in the big endian. for example
  // 000 -> 000
  // 001 -> 100
  // 010 -> 010
  // 011 -> 110
  // 100 -> 001
  // 101 -> 101
  // 110 -> 011
  // 111 -> 111
  // [000, 100, 010, 110, 001, 101, 011, 111]
```